### PR TITLE
ENH: Parse complex number from string

### PR DESCRIPTION
**Pull request by zjpoh**
Addressing #13891. 

I implemented parsing from string to complex number. Here are a couple assumptions that I made
1. If only one component is specified, the other component is assumed to be zero. For example, `'1j'` gives `0+1j` and `'1'` gives `1+0j`.
1. I follow the assumption for other dtypes, where spaces between separator is allowed and will be ignored but spaces in between numbers, `+`, `-`, or `j` are not allowed. For example, `1+1j, 1+1j` is allowed but `1 + 1j,1+1j` or `1+1 j` are not. 

Other comments.
* Clearer documentation on what are the supported dtypes will be helpful.
* Explanation and examples of when things break in the documentation and maybe raising a warning will be helpful. For example, `np.fromstring("1,2,3 4", sep=",")  # array([1, 2, 3])`.

This is my first real numpy PR. Please let me know any advices / comments that you have! Thanks.

### CODE REVIEW

File: numpy/core/src/multiarray/arraytypes.c.src
Comment (charris) [inline numpy/core/src/multiarray/arraytypes.c.src]: The compiler is suggesting an extra set a parenthesis here for clarification, that is why the test failures.

**Comment by rgommers** [comment]
> If only one component is specified, the other component is assumed to be zero.

this is right

>  follow the assumption for other dtypes, where spaces between separator is allowed and will be ignored but spaces in between numbers, +, -, or j are not allowed. 

sounds very reasonable

> Clearer documentation on what are the supported dtypes will be helpful.

yes, would be nice to edit the docstring. also would be good to say in the `dtype` description that complex dtypes are only supported since numpy 1.18.0

> Explanation and examples of when things break in the documentation and maybe raising a warning will be helpful.

not so sure this is necessary, since there's many ways to incorrectly format the strings. one or two examples could be fine, but optional I'd say.

> This is my first real numpy PR. Please let me know any advices / comments that you have! T

From a first read, it looks pretty good. The clear PR description is also nice. Looks like you did everything right:)

File: numpy/core/tests/test_longdouble.py
Comment (rgommers) [inline numpy/core/tests/test_longdouble.py]: bonus points: you could use `assert_raises` and ensure the correct exception kind is given for incorrectly formatted strings

File: numpy/core/src/multiarray/arraytypes.c.src
Comment (mattip) [reply numpy/core/src/multiarray/arraytypes.c.src]: @zipoh ping

Comment (mattip) [reply numpy/core/src/multiarray/arraytypes.c.src]: @zjpoh ping

Comment (zjpoh) [reply numpy/core/src/multiarray/arraytypes.c.src]: Sorry. I'm not available until next Monday. I'll fix it next Monday.

File: numpy/core/tests/test_longdouble.py
Comment (zjpoh) [reply numpy/core/tests/test_longdouble.py]: Similar to the other dtypes, if the string is incorrectly formatted, the function simply reads until it's unable to read. For example,
```python
np.fromstring("1+1", sep=",")  # array([1.])
```
I follow similar pattern and did not check for ill-formatted strings. For example,
```python
np.fromstring("1+1j,2+2", sep=",", dtype="complex")  # array([1.+1.j])
```

**Comment by zjpoh** [comment]
I just notice #13605. I think we should wait until that is merged then I'll make sure that ill-formatted complex number string raise the same error. Then this PR will be ready for review.

**Comment by zjpoh** [comment]
@seberg Thanks for informing! 😄

File: numpy/core/_add_newdocs.py
Comment (seberg) [inline numpy/core/_add_newdocs.py]: Maybe we can just say that most builtin numeric types are supported and extension types may be supported?

File: numpy/core/src/multiarray/arraytypes.c.src
Comment (seberg) [inline numpy/core/src/multiarray/arraytypes.c.src]: This is to indicate an error by not reading everything? I think it could even return a negative value in this case, in any case, I think I would like a comment explaining this path.

File: numpy/core/src/multiarray/arraytypes.c.src
Comment (seberg) [inline numpy/core/src/multiarray/arraytypes.c.src]: Shouldn't this be inside the `*endptr[0] == 'j'` branch? Maybe it does not matter, but that seems clearer to me?

File: numpy/core/tests/test_longdouble.py
Comment (seberg) [reply numpy/core/tests/test_longdouble.py]: What happens if you have `np.fromstring("1j+1")`, also bad formatting I assume?

**Review by seberg — COMMENTED** [review]
If we implement fromtxt, should we also implement the `_scan` functions, so that `fromfile` works as well?

File: numpy/core/src/multiarray/arraytypes.c.src
Comment (zjpoh) [reply numpy/core/src/multiarray/arraytypes.c.src]: Yeap. This is to trigger the error. I'm adding a comment on that. 

I'm fine with changing the imaginary part to -1 but can you explain why returning a negative is preferred since a negative number is equally likely as any other number? Thanks.

File: numpy/core/tests/test_longdouble.py
Comment (zjpoh) [reply numpy/core/tests/test_longdouble.py]: Yeap. I'm assuming that is invalid. What is your thoughts on this? I also added a test case for this. Thanks.

File: numpy/core/src/multiarray/arraytypes.c.src
Comment (seberg) [reply numpy/core/src/multiarray/arraytypes.c.src]: I meant the integer return of the function, would have to check mysefl, but IIRC certain return values may also be accepted to signal error.

File: numpy/core/_add_newdocs.py
Comment (seberg) [inline numpy/core/_add_newdocs.py]: Should probably havea `.. versionadded:` tag.

File: numpy/core/src/multiarray/arraytypes.c.src
Comment (seberg) [inline numpy/core/src/multiarray/arraytypes.c.src]: Silly nit, but can you use
```
/*
 * This is a multiline comment
 * C89 style comment  for multiple lines.
 */
```

File: numpy/core/tests/test_longdouble.py
Comment (seberg) [inline numpy/core/tests/test_longdouble.py]: Tiny Nit: There seems to now be an additional blank line here.

File: numpy/core/tests/test_longdouble.py
Comment (seberg) [inline numpy/core/tests/test_longdouble.py]: Small nits also here: Please make sure to put spaces after all commas in function arguments as per PEP8.

File: numpy/core/tests/test_longdouble.py
Comment (seberg) [inline numpy/core/tests/test_longdouble.py]: Should the last one here have an imaginary part?

**Review by seberg — APPROVED** [review]
Just nitpicking now, the code itself looked nice to me (although I did not have a second super close look). Thanks!

File: numpy/core/src/multiarray/arraytypes.c.src
Comment (seberg) [inline numpy/core/src/multiarray/arraytypes.c.src]: On emore style nit, I think I would remove the empty line here and below (and possibly above)

**Comment by zjpoh** [comment]
@seberg I missed your previous comment on we can add `_scan`. I can take a look at that. I don't mind waiting for that and merge the two PRs together. Thanks~

**Comment by zjpoh** [comment]
I found a function in CPython that parse complex from string https://github.com/python/cpython/blob/1b53a24fb4417c764dd5933bce505f5c94249ca6/Objects/complexobject.c#L784

But I'm not sure what to do with this. Any thoughts?

**Comment by mattip** [comment]
Can we re-use that function? It seems to be designed around convertng an arbitrary python object to a complex, not to be part of a stream-parsing routine. But it would be good to reuse the design of the actual string parsing there, if possible

**Comment by zjpoh** [comment]
Got it. 😃 

I'm not available this week. I'll take a look at it next week.

File: numpy/core/src/multiarray/arraytypes.c.src
Comment (eric-wieser) [inline numpy/core/src/multiarray/arraytypes.c.src]: Can we handle `CLONGDOUBLE` here too? I think we have an `NumPyOS_ascii_strtold` available.

**Pull request merged**
